### PR TITLE
Fix ParticlesMaterial spread

### LINKED ISSUE #47145: ParticlesMaterial's emission direction is wrong if direction.z != 0 and direction.y != 0
**Godot version:**
v3.2.4-rc5

**OS/device including version:**
Windows 10. GLES3.

**Issue description:**
If the direction property of a ParticlesMaterial is changed, it often behaves completely unpredictable.
This is especially apparent, if the values are close to zero.

**Minimal reproduction project:**
The example has two particle systems, which are identical, save for the emission direction.
One of them has the vector (1.0, 0.0, 0.0), the other (1.0, 0.01, 0.01).
They should visually emit in almost the same direction, they are almost 45° off.

[particle_spread_test.zip](https://github.com/godotengine/godot/files/6167382/particle_spread_test.zip)

### PR DESCRIPTION
**Pull request by mortarroad**
Fixes #47145

Instead of doing some butchery with atan and then manipulating angles, we simply rotate the vector.

There is still a singularity if the direction vector is parallel to the local Y axis. This is however only noticeable if the flatness is set to != 0.
This can't be avoided, unless we allow the user to specify a custom up-vector.

### CODE REVIEW

**Comment by QbieShay** [comment]
I'm not sure that Y axis is a good idea to pick as a point of singularity. I see it as being used way more often than other directions (for example a random (0.5,0.2,0.3) (not normalized, I know)) seems like a safer bet to me.

File: scene/resources/particles_material.cpp [hunk 1]
```diff
@@ -319,14 +319,20 @@ void ParticlesMaterial::_update_shader() {
...
 		code += "		direction_yz.z = direction_yz.z / max(0.0001,sqrt(abs(direction_yz.z))); // better uniform distribution\n";
 		code += "		vec3 vec_direction = vec3(direction_xz.x * direction_yz.z, direction_yz.y, direction_xz.z * direction_yz.z);\n";
-		code += "		vec_direction = normalize(vec_direction);\n";
-		code += "		VELOCITY = vec_direction * initial_linear_velocity * mix(1.0, rand_from_seed(alt_seed), initial_linear_velocity_random);\n";
+		code += "		vec3 direction_nrm = normalize(direction);\n";
+		code += "		// rotate spread to direction\n";
+		code += "		vec3 cross_up = vec3(0.0, 1.0, 0.0); // assume local Y is up\n";
+		code += "		vec3 cross_side = cross(cross_up, direction_nrm);\n";
+		code += "		if (length(cross_side) < 0.0001)\n";
+		code += "			cross_side = vec3(0.0, 0.0, 1.0);\n";
```
Comment (clayjohn) [inline scene/resources/particles_material.cpp:331]:
```suggestion
		code += "		if (length(cross_side) < 0.0001) {\n";
		code += "			cross_side = vec3(0.0, 0.0, 1.0);\n";
		code += "		}\n";
```

We try to always use braces, even for single statements. This helps enforce best practices for beginners. 

https://docs.godotengine.org/en/latest/tutorials/shaders/shaders_style_guide.html#line-breaks-and-blank-lines

**Review by clayjohn — COMMENTED** [review]
Testing this out locally and it looks great. Subject to my one code style comment I think we should merge it. 

@QbieShay I think if the singularity ends up being noticeable for a lot of people we should indeed add a way to rotate the "up" vector. Making it a random angle may help for some users, but it also just masks the problem in an unintuitive way.

**Comment by clayjohn** [comment]
Also, this PR needs to target 4.0 and then be cherry picked for 3.x. 

Let me know if you need help rebasing this on the 4.0 branch.

**Comment by QbieShay** [comment]
@clayjohn I think that having a singularity that happens in one of the most common use cases is not a good approach

**Pull request closed without merging**